python/revng/cli/_commands/idb_converter.py

```diff
@@ -0,0 +1,655 @@
+            n_elements = type.type_details.n_elems
+            if n_elements == 0:
+                logger.warning(f"Array {type_name} has invalid zero size")
+            qualified_type.Qualifiers.insert(
+                0, m.Qualifier(Kind=m.QualifierKind.Array, Size=n_elements)
+            )
+
+        if type.is_decl_const():
+            qualified_type.Qualifiers.insert(0, CONST_QUALIFIER)
+
+        if ordinal is not None:
+            self.idb_types_to_revng_types[ordinal] = qualified_type
+        self.revng_types_by_id[revng_type.ID] = revng_type
+
+        return qualified_type
+
+    def _get_primitive_type(self, kind: m.PrimitiveTypeKind, size: int) -> m.QualifiedType:
+        """Gets a primitive type, taking care to register it"""
+        revng_type = m.PrimitiveType(PrimitiveKind=kind, Size=size)
+        if revng_type.ID not in self.revng_types_by_id:
+            self.revng_types_by_id[revng_type.ID] = revng_type
+
+        return m.QualifiedType(UnqualifiedType=m.Reference.create(m.Binary, revng_type))
+
+    def get_model(self) -> m.Binary:
+        return m.Binary(
+            # NOTE: We assume that the EntryPoint can be obtained from binary itself, so we use an
+            # invalid address for it here.
+            EntryPoint=m.MetaAddress(Address=0x0, Type=MetaAddressType.Invalid),
+            Functions=list(self.functions),
+            ImportedDynamicFunctions=list(self.dynamic_functions),
+            Types=list(self.revng_types_by_id.values()),
+            Architecture=self.arch,
+            Segments=self.segments,
+            ImportedLibraries=self.imported_libraries,
+        )
+
+    def resolve_typeref(self, typeref: m.Reference) -> Optional[RevngTypes]:
+        return self.revng_types_by_id.get(typeref.id)
+
+    def get_revng_type_by_name(self, name):
+        for revng_type in self.revng_types_by_id.values():
+            if revng_type.OriginalName == name:
+                return revng_type
+        return None
+
+    def unwrap_qualified(self, qt: m.QualifiedType):
+        if qt.Qualifiers:
+            raise ValueError("Trying to unwrap qualified type with non empty qualifiers list!")
+
+        return self.resolve_typeref(qt.UnqualifiedType)
+
+
+def get_primitive_kind(idb_type: idb.typeinf.TInfo) -> m.PrimitiveTypeKind:
+    if idb_type.is_decl_void():
+        return m.PrimitiveTypeKind.Void
+    elif idb_type.is_decl_floating():
+        return m.PrimitiveTypeKind.Float
+    elif idb.typeinf_flags.is_type_integral(idb_type.get_decltype()):
+        if idb_type.is_signed():
+            return m.PrimitiveTypeKind.Signed
+        elif idb_type.is_unsigned():
+            return m.PrimitiveTypeKind.Unsigned
+        else:
+            return m.PrimitiveTypeKind.Number
+    elif idb_type.is_decl_ptr():
+        # TODO: can we ever reach this point?
+        return m.PrimitiveTypeKind.PointerOrNumber
+    elif idb_type.is_decl_bitfield():
+        return m.PrimitiveTypeKind.Generic
```

Comment:
IIRC we don't support bitfields, we should ignore/mark the types containing them as invalid.